File: src/policy/nat/nat_rule.rs
use ipnet::IpNet;

use crate::policy::nat::port_range::PortRange;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum NatProtocol {
    Tcp,
    Udp,
    Icmp,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NatAction {
    Pat,
    Dnat,
    Snat,
    Masquerade,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NatRule {
    id: String,
    priority: u32,
    in_interface: Option<String>,
    out_interface: Option<String>,
    in_zone: Option<String>,
    out_zone: Option<String>,
    src_cidr: Option<IpNet>,
    dst_cidr: Option<IpNet>,
    protocol: Option<NatProtocol>,
    src_port: Option<u16>,
    dst_port: Option<u16>,
    action: NatAction,
}

impl NatRule {
    pub fn new(
        id: String,
        priority: u32,
        in_interface: Option<String>,
        out_interface: Option<String>,
        in_zone: Option<String>,
        out_zone: Option<String>,
        src_cidr: Option<IpNet>,
        dst_cidr: Option<IpNet>,
        protocol: Option<NatProtocol>,
        src_port: Option<u16>,
        dst_port: Option<u16>,
        action: NatAction,
    ) -> Self {
        Self {
            id,
            priority,
            in_interface,
            out_interface,
            in_zone,
            out_zone,
            src_cidr,
            dst_cidr,
            protocol,
            src_port,
            dst_port,
            action,
        }
    }

    pub fn id(&self) -> &str {
        &self.id
    }
    pub fn priority(&self) -> u32 {
        self.priority
    }
    pub fn in_interface(&self) -> Option<&str> {
        self.in_interface.as_ref().map(|s| s.as_str())
    }
    pub fn out_interface(&self) -> Option<&str> {
        self.out_interface.as_ref().map(|s| s.as_str())
    }
    pub fn in_zone(&self) -> Option<&str> {
        self.in_zone.as_ref().map(|s| s.as_str())
    }
    pub fn out_zone(&self) -> Option<&str> {
        self.out_zone.as_ref().map(|s| s.as_str())
    }
    pub fn src_cidr(&self) -> Option<IpNet> {
        self.src_cidr
    }
    pub fn dst_cidr(&self) -> Option<IpNet> {
        self.dst_cidr
    }
    pub fn protocol(&self) -> Option<NatProtocol> {
        self.protocol
    }
    pub fn src_port(&self) -> Option<u16> {
        self.src_port
    }
    pub fn dst_port(&self) -> Option<u16> {
        self.dst_port
    }
    pub fn action(&self) -> NatAction {
        self.action.clone()
    }
}

File: frontend/src/types/nat/NatRule.ts
export type NatActionKind = "snat" | "dnat" | "pat" | "masquerade";
export type NatType = "SNAT" | "DNAT" | "PAT" | "MASQUERADE";

export type NatProtocolName =
  | "NAT_PROTOCOL_ALL"
  | "NAT_PROTOCOL_TCP"
  | "NAT_PROTOCOL_UDP"
  | "NAT_PROTOCOL_ICMP";

export type NatProtocolCode = 0 | 1 | 2 | 3 | -1;
export type NatProtocol = NatProtocolName | NatProtocolCode;

export type NatRuleAction =
  | {
      $case: "snat";
      snat: {
        srcCidr: string;
        translatedIp: string;
        srcPortMin?: number | null;
        srcPortMax?: number | null;
      };
    }
  | { $case: "dnat"; dnat: { dstCidr: string; translatedIp: string } }
  | {
      $case: "pat";
      pat: {
        dstIp: string;
        dstPort: number;
        translatedIp: string;
        translatedPort: number;
      };
    }
  | {
      $case: "masquerade";
      masquerade: {
        srcCidr?: string | null;
        srcPortMin?: number | null;
        srcPortMax?: number | null;
      };
    };

export interface NatRule {
  id: string;
  isActive: boolean;
  priority: number;
  protocol: NatProtocol;
  inInterface: string | null;
  outInterface: string | null;
  inZone: string | null;
  outZone: string | null;
  matchSrcPortMin: number | null;
  matchSrcPortMax: number | null;
  matchDstPortMin: number | null;
  matchDstPortMax: number | null;
  action: NatRuleAction;
  createdAt: string;
  updatedAt: string;
}

export type NatRuleUpsertBody = {
  isActive: boolean;
  priority: number;
  protocol: NatProtocolName;
  inInterface?: string | null;
  outInterface?: string | null;
  inZone?: string | null;
  outZone?: string | null;
  matchSrcPortMin?: number | null;
  matchSrcPortMax?: number | null;
  matchDstPortMin?: number | null;
  matchDstPortMax?: number | null;
  action: NatRuleAction;
};

export function getNatRuleType(rule: NatRule): NatType {
  return actionKindToType(rule.action.$case);
}

export function actionKindToType(kind: NatActionKind): NatType {
  switch (kind) {
    case "snat":
      return "SNAT";
    case "dnat":
      return "DNAT";
    case "pat":
      return "PAT";
    case "masquerade":
      return "MASQUERADE";
  }
}

export function typeToActionKind(type: NatType): NatActionKind {
  switch (type) {
    case "SNAT":
      return "snat";
    case "DNAT":
      return "dnat";
    case "PAT":
      return "pat";
    case "MASQUERADE":
      return "masquerade";
  }
}

export function protocolLabel(protocol: NatProtocol): string {
  switch (protocol) {
    case 0:
    case "NAT_PROTOCOL_ALL":
      return "ALL";
    case 1:
    case "NAT_PROTOCOL_TCP":
      return "TCP";
    case 2:
    case "NAT_PROTOCOL_UDP":
      return "UDP";
    case 3:
    case "NAT_PROTOCOL_ICMP":
      return "ICMP";
    default:
      return "UNKNOWN";
  }
}
